pennylane/debugging/operation_checker.py

```diff
@@ -0,0 +1,833 @@
+    matrix_from_matrix,
+    matrix_from_sparse_matrix,
+    matrix_from_terms,
+    matrix_from_decomposition,
+    matrix_from_generator,
+]
+
+
+class CheckerError(Exception):
+    """An internal error raised in OperationChecker used to mark specific exceptions."""
+
+
+class OperationChecker:
+    r"""Check one or multiple operation subclasses to define all required properties,
+    be well-defined, and have consistent properties.
+
+    Args:
+        verbosity (str): How much output to print during execution (also see below):
+
+            - ``"pass"``: Print all errors, hints, comments and status reports;
+
+            - ``"comment"``: Like ``"pass"`` but without status reports;
+
+            - ``"hint"``: Only print errors and hints;
+
+            - ``"error"``: Only print errors.
+
+        max_num_params (int): Largest number of parameters to check for operations
+            that do not provide a fixed number of parameters via ``num_params`` themselves.
+        print_color (bool): Whether or not to use colors in the terminal and returned outputs.
+        tol (float): Numeric (absolute) tolerance for comparing matrices.
+
+    The categorization of test results and of the associated messages is as follows:
+
+    - ``"pass"``: Status reports e.g. after a successfully completed run.
+
+    - ``"comment"``: Comments regarding certain properties of the tested operation(s).
+      These does not require any action to change the operation but is used to raise awareness
+      for behaviour that might be unexpected or differing from common operations.
+
+    - ``"hint"``: Remarks similar to warnings that indicate concrete hints to change the checked
+      operation(s). The recommended changes are expected to improve the code quality, performance
+      or consistency with other operations.
+
+    - ``"error"``: Problems with the checked operation(s) that require changes. These problems
+      might be in the core of the ``Operation``, preventing instantiation, or in a specific
+      method or property that is rendered unusable or inconsistent by the problem.
+    """
+
+    def __init__(self, verbosity="pass", max_num_params=10, print_color=True, tol=1e-5):
+        # pylint: disable=too-many-instance-attributes
+        self._verbosity = {
+            key for key, val in verbosity_levels.items() if val <= verbosity_levels[verbosity]
+        }
+        self.max_num_params = max_num_params
+        self.print_color = print_color
+        self.tol = tol
+        self.results = self.output = self.tmp = self.seed = None
+
+    def __call__(self, op, parameters=None, wires=None, seed=None):
+        r"""Call the OperationChecker on one or multiple operations.
+
+        Args:
+            op (type or .operation.Operation): Operation(s) to check, either provided as class
+                or as instance. Allowed to be a ``Sequence`` of classes or instances instead,
+                in which case the function iterates over all objects.
```

Comment:
If it is allowed to be a sequence, best to add this to the type info :)

Having said that, I feel like this may be a complication - why not just allow users to pass a single op and loop over the call?